File: ManagedSpyLib/EventTargetWindow.cpp
#include "StdAfx.h"
#include "Commands.h"
#include "EventTargetWindow.h"

using namespace System;
using namespace System::Collections::Generic;

void EventTargetWindow::WndProc(Message% m) {

	if (m.Msg == WM_EVENTFIRED) {
		//an event has been fired. 

		MemoryStore* store = MemoryStore::OpenStore((int)m.WParam, (int)m.LParam, true);
		if (store!= NULL) {
			List<Object^>^ params= (List<Object^>^)store->GetParameters();
			if (params != nullptr) {
				_ASSERTE(params->Count == 3);
				if (Desktop::proxyCache->ContainsKey((IntPtr)params[0])) {
					ControlProxy^ proxy = Desktop::proxyCache[(IntPtr)params[0]];
					_ASSERTE(proxy != nullptr);
					//fire the event.
					EventDescriptorCollection^ events = proxy->GetEvents();
					if (events != nullptr && (int)params[1]<= events->Count && (int)params[1]>=0) {
						proxy->RaiseEvent(gcnew ProxyEventArgs(events[(int)params[1]], (EventArgs^)params[2]));
					}
				}
			}
			store->Release();	//release, don't delete if in the same process.
		}
	}
	else if (m.Msg == WM_WINDOWDESTROYED) {
		if (Desktop::proxyCache->ContainsKey(m.WParam)) {
			Desktop::proxyCache->Remove(m.WParam);
		}
	}
	else if (m.Msg == WM_HANDLECHANGED) {
		IntPtr oldHandle = m.WParam;
		IntPtr newHandle = m.LParam;
		ControlProxy^ proxy = nullptr;
		if (Desktop::proxyCache->ContainsKey(oldHandle)) {
			proxy = Desktop::proxyCache[oldHandle];
		}

		if (proxy == nullptr) {
			for each (KeyValuePair<IntPtr, ControlProxy^> entry in Desktop::proxyCache) {
				if (entry.Value != nullptr && entry.Value->Handle == oldHandle) {
					proxy = entry.Value;
					break;
				}
			}
		}

		if (proxy != nullptr) {
			List<IntPtr>^ staleKeys = gcnew List<IntPtr>();
			for each (KeyValuePair<IntPtr, ControlProxy^> entry in Desktop::proxyCache) {
				if (Object::ReferenceEquals(entry.Value, proxy) || entry.Key == oldHandle || entry.Key == newHandle) {
					staleKeys->Add(entry.Key);
				}
			}

			for each (IntPtr key in staleKeys) {
				Desktop::proxyCache->Remove(key);
			}

			proxy->Handle = newHandle;
			if (newHandle != IntPtr::Zero && !Desktop::proxyCache->ContainsKey(newHandle)) {
				Desktop::proxyCache->Add(newHandle, proxy);
			}
		}
	}
	else {
		Control::WndProc(m);
	}
}

File: ManagedSpyLib/EventTargetWindow.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Windows.Forms;

namespace Microsoft.ManagedSpy
{
    public static class EventDispatchHelper
    {
        public static bool TryGetEventDescriptor(
            EventDescriptorCollection events,
            int eventIndex,
            out EventDescriptor eventDescriptor)
        {
            eventDescriptor = null;
            if (events == null || eventIndex < 0 || eventIndex >= events.Count)
            {
                return false;
            }

            eventDescriptor = events[eventIndex];
            return eventDescriptor != null;
        }
    }

    internal sealed class EventTargetWindow : Control
    {
        protected override void WndProc(ref Message m)
        {
            uint messageId = unchecked((uint)m.Msg);
            if (messageId == ManagedSpyMessages.EventFired)
            {
                MemoryStore store = MemoryStore.OpenStore((int)m.WParam, (int)m.LParam, true);
                if (store != null)
                {
                    try
                    {
                        List<object> parameters = store.GetParameters() as List<object>;
                        if (parameters != null && parameters.Count == 3)
                        {
                            IntPtr proxyHandle = (IntPtr)parameters[0];
                            if (Desktop.ProxyCache.TryGetValue(proxyHandle, out ControlProxy proxy))
                            {
                                EventDescriptorCollection events = proxy.GetEvents();
                                if (EventDispatchHelper.TryGetEventDescriptor(events, (int)parameters[1], out EventDescriptor eventDescriptor))
                                {
                                    proxy.RaiseEvent(new ProxyEventArgs(eventDescriptor, (EventArgs)parameters[2]));
                                }
                            }
                        }
                    }
                    finally
                    {
                        store.Release();
                    }
                }

                return;
            }

            if (messageId == ManagedSpyMessages.WindowDestroyed)
            {
                Desktop.ProxyCache.Remove(m.WParam);
                return;
            }

            if (messageId == ManagedSpyMessages.HandleChanged)
            {
                IntPtr oldHandle = m.WParam;
                IntPtr newHandle = m.LParam;
                if (Desktop.ProxyCache.TryGetValue(oldHandle, out ControlProxy proxy))
                {
                    Desktop.ProxyCache.Remove(oldHandle);
                    proxy.Handle = newHandle;
                    if (newHandle != IntPtr.Zero)
                    {
                        Desktop.ProxyCache[newHandle] = proxy;
                    }
                }

                return;
            }

            base.WndProc(ref m);
        }
    }
}
